src/session/user.rs

```diff
@@ -146,8 +147,65 @@ mod imp {
 
             let avatar = obj.avatar();
             let user_expression = gtk::ConstantExpression::new(obj);
-            let full_name_expression = super::User::full_name_expression(&user_expression);
-            full_name_expression.bind(avatar, "display-name", gtk::NONE_WIDGET);
+            let display_name_expression = super::User::display_name_expression(&user_expression);
+            display_name_expression.bind(avatar, "display-name", gtk::NONE_WIDGET);
+        }
+    }
+    pub fn time_ago(unix: i32) -> String {
+        let datetime_utc = glib::DateTime::new_now_utc().unwrap().to_unix();
+        let unix: i64 = unix.into();
+        let last_seen = glib::DateTime::from_unix_utc(unix)
+            .unwrap()
+            .to_local()
+            .unwrap();
+        let datetime_local = glib::DateTime::new_now_local().unwrap();
+
+        let diff = datetime_utc - unix;
+        let minutes = diff / 60;
+        let hours = minutes / 60;
+        if datetime_local.format("%Y %m %d").unwrap() == last_seen.format("%Y %m %d").unwrap() {
+            if hours == 0 {
+                if minutes < 0 {
+                    gettext("just now")
+                } else if minutes == 1 {
+                    gettext("a minute ago")
+                } else {
+                    format!("{} {}", gettext("minutes ago"), minutes)
+                }
+            } else if hours == 1 {
+                gettext("an hour ago")
+            } else {
+                format!("{} {}", gettext("hours ago"), hours)
+            }
+        } else if datetime_local.format("%Y %m %d").unwrap()
+            == last_seen.add_days(1).unwrap().format("%Y %m %d").unwrap()
+        {
+            last_seen
+                .format(&gettext("yesterday at %l:%M %p"))
+                .unwrap()
+                .to_string()
+        } else if datetime_local.format("%Y %W").unwrap() == last_seen.format("%Y %W").unwrap() {
+            last_seen
+                .format(&gettext("on %A at %l:%M %p"))
+                .unwrap()
+                .to_string()
+        } else if datetime_local.format("%Y %W").unwrap()
+            == last_seen.add_weeks(1).unwrap().format("%Y %W").unwrap()
+        {
+            last_seen
+                .format(&gettext("last %A at %l:%M %p"))
+                .unwrap()
+                .to_string()
+        } else if datetime_local.format("%Y").unwrap() == last_seen.format("%Y").unwrap() {
+            last_seen
+                .format(&gettext("on %d %b at %l:%M %p"))
+                .unwrap()
+                .to_string()
+        } else {
+            last_seen
+                .format(&gettext("%d.%m.%Y at %l:%M %p"))
+                .unwrap()
+                .to_string()
```

Comment:
I put this conversation at the end. What I dislike about this function is that you rely entirely on string comparisons.

The function could look like much cleaner like this
```rust
    pub fn time_ago(unix: i32) -> String {
        let datetime_now = glib::DateTime::new_now_local().unwrap();
        let last_seen = glib::DateTime::from_unix_utc(unix as i64)
            .unwrap()
            .to_local()
            .unwrap();

        let diff = datetime_now.difference(&last_seen);

        let seconds = diff / 1000000;
        let minutes = seconds / 60;
        let hours = minutes / 60;
        let days = hours / 24;

        if days == 0 {
            if hours == 0 {
                if minutes == 0 {
                    gettext("just now")
                } else if minutes == 1 {
                    gettext("a minute ago")
                } else {
                    format!("{} {}", gettext("minutes ago"), minutes)
                }
            } else if hours == 1 {
                gettext("an hour ago")
            } else {
                format!("{} {}", gettext("hours ago"), hours)
            }
        } else if days == 1 {
            last_seen
                .format(&gettext("yesterday at %l:%M %p"))
                .unwrap()
                .to_string()
        } else if days < 7 {
            last_seen
                .format(&gettext("on %A at %l:%M %p"))
                .unwrap()
                .to_string()
        } else if days < 14 {
            last_seen
                .format(&gettext("last %A at %l:%M %p"))
                .unwrap()
                .to_string()
        } else if datetime_now.year() == last_seen.year() {
            last_seen
                .format(&gettext("on %d %b at %l:%M %p"))
                .unwrap()
                .to_string()
        } else {
            last_seen
                .format(&gettext("%d.%m.%Y at %l:%M %p"))
                .unwrap()
                .to_string()
        }
    }
```
I haven't tested this, but I think such an approach is the correct one.

Reply:
> `format!("{} {}", gettext("hours ago"), hours)` 
my bad it should be like

`format!("{} {}", hours, gettext("hours ago"),)`

Reply:
` else if days == 1` I also was thinking about it like that, but calendar doesn't work like that, if you send a message at 23:59, one minute later there'll be new day, but the function will be still be printing `1 minute ago`

Reply:
> ` else if days == 1` I also was thinking about it like that at the beggining, but calendar doesn't work like that. If you send a message at 23:59, one minute later there'll be a new day, but the function will be still be printing `1 minute ago`

I hope I understood you correctly:
I am not sure what the official client does, but printing `1 minute ago` if it's currently 00:00 and the user was last online on the previous day at 23:59 is what I as a user would expect.
The "time_ago" message should be consistent all around the clock, no matter if there is a new day in the meantime.
IMO, this means it should be relative to the current time because in this way it doesn't violate the principle of least surprise and is also easier to implement.

Reply:
But even if we don't want it like I described, we still should get rid of the string comparisons.
We could do this instead:
```rust
if (datetime_now.year(), datetime_now.month(), datetime_now.day_of_month()) == (last_seen.year(), last_seen.month(), last_seen.day_of_month()) {
    // bla
}
```

Reply:
I think @melix99 will have the last word here.

Reply:
Oops, didn't notice the notification. I think checking the time difference is definitely what we want (so @marhkb's first comment).

Reply:
Ok, I've made a few changes to the function to find a compromise Please share your thoughts about it
https://gist.github.com/just-carlod/cce4e2c5a1a772eaec7433b59e22ec79

Reply:
A lot better. Can you force push that so that we can discuss this further?

Reply:
sure, I'll do it soon

Reply:
https://github.com/melix99/telegrand/pull/203#discussion_r782284356
https://github.com/melix99/telegrand/pull/203#discussion_r782287365
I replied to wrong comment lol